### PR TITLE
Add an admin option to shared secret registration (breaks backwards compat)

### PR DESCRIPTION
**Pull request by erikjohnston**

### CODE REVIEW

File: synapse/rest/client/v1/register.py [hunk 1]
```diff
@@ -324,24 +324,28 @@ def _do_shared_secret(self, request, register_json, session):
...
             key=self.hs.config.registration_shared_secret,
-            msg=user,
             digestmod=sha1,
-        ).hexdigest()
-
-        password = register_json["password"].encode("utf-8")
+        )
+        want_mac.update(user)
+        want_mac.update(password)
+        want_mac.update("admin" if admin else "notadmin")
```
Comment (NegativeMjark) [inline synapse/rest/client/v1/register.py]: What happens if someone creates a user with password = "password" and admin=False and then someone nicks their mac and registers a user with password = "passwordnot" and admin=True?

Comment (NegativeMjark) [inline synapse/rest/client/v1/register.py]: Similar thoughts on shifting characters between the "user" and "password" fields.

Comment (erikjohnston) [inline synapse/rest/client/v1/register.py]: Yeah, leo just brought that up in the sytest PR. I've added null separators, which I think makes it fine?

**Comment by NegativeMjark** [comment]
Looks like it might work.

**Pull request merged**